### ISSUE TITLE
Review for JOSS Submission

### ISSUE THREAD

**Issue by avakiai** [rempsyc/rempsyc#16]
# Review
Overall, I think the package is well written, very well documented, and offers a sufficiently complete set of convenience functions that can adequately meet the needs of some researchers who are relatively unfamiliar with R. You may find my review checklist [here](https://github.com/openjournals/joss-reviews/issues/5466#issuecomment-1596115609).

### 1. Dependency issues
As this package wraps convenience functions around existing packages, solid dependency handling is key.

Both at the point of installation and when certain functions are called, I encountered numerous namespace errors. I think the main issue is that this package relies on certain (newer) versions of depencency packages, but package functions only check whether the package is installed or not. I would recommend adding the minimum package requirement in the `version` argument in calls to `rlang::check_installed()`.

Example of what happens if you try to use a function where a dependency is not installed at all:
```
> nice_table(t.tests)
ℹ The package "flextable" is required for this function.
✖ Would you like to install it?

1: Yes
2: No
```

Example of what happens if you try to use a function where a dependency is installed, but a newer version is required:
```
> moderations <- nice_mod(
+   data = mtcars,
+   response = c("mpg", "disp"),
+   predictor = "gear",
+   moderator = "wt"
+ )
Error: 'r2_semipartial' is not an exported object from 'namespace:effectsize'
```
The above was solved by updating the `effectsize` package with `install.packages("effectsize")`.

This issue therefore affects two points in the checklist:

#### i. Functionality/Installation
Installation via `install.packages("rempsyc")` resolves without issue, but calling `library(rempsyc)` throws error:
`Error: package or namespace load failed for ‘rempsyc’:
 object ‘pick’ is not exported by 'namespace:dplyr'`

Once I updated `dplyr` from 1.0.9 to 1.1.2, then package loaded correctly. Thus, minimum version as stated on CRAN should be updated.

#### ii. Documentation/Installation instructions
Dependencies and required versions are not stated on the main documentation page, though they are stated on CRAN. I'm not sure what is considered best practice for loading depencies, but the author may consider including a step for the manual installation of all depencies or inserting a chunk for installing dependencies in the "Getting Started" section of each tutorial. rlang dependecy checks could then be kept as a backup. Indeed, it feels a bit hacky to interrupt workflow for installing a dependency. 

#### i. Documentation/Automated tests
testthat only available for `nice_violin()` currently. Should be implemented for all functions (?).

## Minor Points & Suggestions
1. The tutorial for the `nice_mod()` may benefit from a little bit more explanation or perhaps link to a couple resources for setting up a moderation/simple slopes analysis. It may be obvious, but I would also suggest explicitly mentioning that these functions compute straight linear models under the hood. 

2. It may be useful to add an argument that would allow a user to get row indices from `extract_duplicates()` and `best_duplicate()` that would make data wrangling easier. 

3. I would suggest including a chunk in the tutorial for [`nice_randomization()`](https://rempsyc.remi-theriault.com/articles/randomize.html) encouraging users to set (for reproducibility) or randomize the seed before running the function. 

4. Users may wish to use the venn diagram visualization for applications other than IOS. Thus, it might be useful if the maximum overlap for `overlap_circle()` could go up to 100% (it currently only goes up to 85%).  Relatedly, as some users may wish to use the venn diagram with inputs other than (I presume) the 1-7 IOS scale, you may consider allowing users to directly define the percent overlap (e.g. 57%) and/or provide the values in each section explicitly (e.g. c(10,27,10)). Admittedly, this may make the function deviate from its intended purpose, and for such uses users could directly use functions from other packages.

5. Print color on warning messages (e.g. output from `nice_t_test()`) is uses the function `message_white()` which is very difficult to read on the default white background of R. I would suggest changing to the default message function.

6. It is a bit unclear to me how (or whether) `nice_contrasts()` can handle nested comparisons for marginal means. In the syntax of *emmeans*, it is straightforward to define whether you want to know contrasts for levels of var1 within var2, and vice versa.

**Comment by rempsyc** [comment]
Dear Ava,

First, a sincere thank you for this very helpful and constructive review. You will find my answers below:

> I think the main issue is that this package relies on certain (newer) versions of depencency packages, but package functions only check whether the package is installed or not. I would recommend adding the minimum package requirement in the version argument in calls to rlang::check_installed().

Thank you for highlighting this very important point that I have missed so far. I believed that `rlang::check_installed()` checked the versions from the DESCRIPTION file, but I was wrong. Thus, I have added the proper required versions in `rlang::check_installed()` in CRAN version 0.1.3. However, in order to avoid having to update version numbers at two different places (which increases maintenance and risk of errors), I have also developed a new function to handle this, `get_dep_version()` (in dev version 0.1.3.1). Example usage:

``` r
rempsyc::get_dep_version(dep = "flextable", pkg = "rempsyc")
#> [1] "0.9.1"
rempsyc::get_dep_version(dep = "effectsize", pkg = "rempsyc")
#> [1] "0.8.3"
```

<sup>Created on 2023-06-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

> Installation via install.packages("rempsyc") resolves without issue, but calling library(rempsyc) throws error:
Error: package or namespace load failed for ‘rempsyc’: object ‘pick’ is not exported by 'namespace:dplyr'

Thanks, this has been resolved in #14 (as of development version 0.1.2.1, and CRAN version 0.1.3).

> Dependencies and required versions are not stated on the main documentation page, though they are stated on CRAN. I'm not sure what is considered best practice for loading depencies, but the author may consider including a step for the manual installation of all depencies or inserting a chunk for installing dependencies in the "Getting Started" section of each tutorial. rlang dependecy checks could then be kept as a backup. Indeed, it feels a bit hacky to interrupt workflow for installing a dependency.

Thanks, I have now updated the README file to include a section on dependencies explaining the minimalist dependency philosophy, pointing to CRAN for the full list, and showing a way to install all suggested dependencies at once (below; as of version 0.1.3.1).

```` r
**Dependencies:** Because `rempsyc` is a package of convenience functions relying on several external packages, 
it uses (inspired by the [`easystats`](https://easystats.github.io/easystats/) packages) a minimalist philosophy of 
only installing packages that you need when you need them through `rlang::check_installed()`. Should you wish to 
specifically install all suggested dependencies at once (you can view the full list by clicking on the CRAN badge 
on this page), you can run the following (be warned that this may take a long time, as some of the suggested 
packages are only used in the vignettes or examples):
``` r
install.packages("rempsyc", dependencies = TRUE)
```
````

In version 0.1.3.2, I have also inserted a chunk for installing dependencies in the "Getting Started" section of each tutorial, and created a new function for this purpose, `install_if_not_installed`:

```` r
> ***Note:*** If you haven't installed this package yet, you will need to install it via the following command: 
`install.packages("rempsyc")`. Furthermore, you may be asked to install the following packages if you haven't 
installed them already (you may decide to install them all now to avoid interrupting your workflow if you wish 
to follow this tutorial from beginning to end):

```{r}
pkgs <- c("flextable", "performance", "see", "lmtest", "ggplot2",
          "qqplotr", "ggrepel", "patchwork", "boot")
install_if_not_installed(pkgs)
```
````

> testthat only available for nice_violin() currently. Should be implemented for all functions (?).

Thanks, I have now added tests for all functions (version 0.1.3.1).

> The tutorial for the nice_mod() may benefit from a little bit more explanation or perhaps link to a couple resources for setting up a moderation/simple slopes analysis. It may be obvious, but I would also suggest explicitly mentioning that these functions compute straight linear models under the hood.

Thanks, I have added the following in the [vignette](https://rempsyc.remi-theriault.com/articles/moderation.html) (version 0.1.3.1):

``` r
Sometimes in research, we want to know whether the effect of variable X on Y is affected by a third variable, 
variable Z. In other terms, we ask if there is an *interaction* between variables X and Z, and their effects on Z.

![](https://rempsyc.remi-theriault.com/images//blog_mod/moderation.png)

Note that this is different from *mediation*, where the mediator is the mechanism that *explains* the link 
between X and Y (rather than a variable that *modifies* an existing relationship like in moderation).

![](https://rempsyc.remi-theriault.com/images//blog_mod/mediation.png)

In R, we conduct **moderation** analyses using straight linear models with the `lm` function, and we specify 
interaction effects with the * operator. Not everyone is familiar with using `lm` however, so `rempsyc` provides 
a (relatively) simpler interface where it is straightforward what variable is the moderator, and which one is the 
predictor. Although it does not make a difference between the `lm` model, for some (e.g., that find the `lm` 
function scary), it can be helpful to think about these variables in this way. The other benefit is that it provides 
a useful effect size and its 95% confidence interval, and formats everything in a table ready to be exported to 
word through `nice_table`.

The topic of moderations and simple slopes can be a complex one. It is not the goal of this tutorial to describe 
the theory behind it, only to show a practical way to do them. For a more detailed reading on the topic, please 
see one of the existing excellent sources on the topic 
([1](https://mspeekenbrink.github.io/sdam-r-companion/moderation-and-mediation.html), 
[2](https://bookdown.org/pingapang9/linear_models_bookdown/chap-moderation.html), 
[3](https://ademos.people.uic.edu/Chapter14.html)).
```
And in the simple slopes section:

```r
You might have heard about "simple slopes" before. But what does that mean? Essentially, this means looking at the 
strength (regression coefficient) and significance of the slope, when subsetting for observations that are high, low, 
or average on a variable, typically the moderating variable. A bit further down, this will get clearer by looking 
at the plot of the interaction, which shows one slope for observations that are high on the wt (moderating) 
variable, a second slope for those that are low, and a third slope for those that are average.
```

> It may be useful to add an argument that would allow a user to get row indices from extract_duplicates() and best_duplicate() that would make data wrangling easier.

Thanks. `extract_duplicates` already provides row indices by default in the “Row” column:

``` r
df1 <- data.frame(
  id = c(1, 2, 3, 1, 3),
  item1 = c(NA, 1, 1, 2, 3),
  item2 = c(NA, 1, 1, 2, 3),
  item3 = c(NA, 1, 1, 2, 3)
)

rempsyc::extract_duplicates(df1, id = "id")
#>   Row id item1 item2 item3 count_na
#> 1   1  1    NA    NA    NA        3
#> 2   4  1     2     2     2        0
#> 3   3  3     1     1     1        0
#> 4   5  3     3     3     3        0
```

<sup>Created on 2023-06-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

`best_duplicate` does not because it is typically used on the original data frame, which usually we want to avoid modifying. That said, I added an argument, `keep.rows`, to do so (defaults to FALSE; version 0.1.3.2):

``` r
df1 <- data.frame(
  id = c(1, 2, 3, 1, 3),
  item1 = c(NA, 1, 1, 2, 3),
  item2 = c(NA, 1, 1, 2, 3),
  item3 = c(NA, 1, 1, 2, 3)
)

rempsyc::best_duplicate(df1, id = "id", keep.rows = TRUE)
#> (2 duplicates removed)
#>   Row id item1 item2 item3
#> 1   4  1     2     2     2
#> 2   2  2     1     1     1
#> 3   3  3     1     1     1
```

<sup>Created on 2023-06-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

> I would suggest including a chunk in the tutorial for nice_randomization() encouraging users to set (for reproducibility) or randomize the seed before running the function.

Thanks, [added](https://rempsyc.remi-theriault.com/articles/randomize.html) (in dev version 0.1.2.5, and CRAN version 0.1.3).

> Users may wish to use the venn diagram visualization for applications other than IOS. Thus, it might be useful if the maximum overlap for overlap_circle() could go up to 100% (it currently only goes up to 85%). Relatedly, as some users may wish to use the venn diagram with inputs other than (I presume) the 1-7 IOS scale, you may consider allowing users to directly define the percent overlap (e.g. 57%) and/or provide the values in each section explicitly (e.g. c(10,27,10)). Admittedly, this may make the function deviate from its intended purpose, and for such uses users could directly use functions from other packages.

Thanks, added (in dev version 0.1.3.1). Reprex:

``` r
rempsyc::overlap_circle(55, scoring = "percentage")
```

![](https://i.imgur.com/kHxePoZ.png)

``` r

rempsyc::overlap_circle(100, scoring = "percentage")
```

![](https://i.imgur.com/t5UHKi2.png)

``` r

rempsyc::overlap_circle(c(17, 10, 97), scoring = "direct",
                        categories = c("Self", "Outgroup"))
```

![](https://i.imgur.com/xmqQM9f.png)

<sup>Created on 2023-06-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

The [vignette](https://rempsyc.remi-theriault.com/articles/circles.html) has also been updated to reflect this new change.

> Print color on warning messages (e.g. output from nice_t_test()) is uses the function message_white() which is very difficult to read on the default white background of R. I would suggest changing to the default message function.

Thanks! I had not noticed because I use a dark Rstudio theme… This has been corrected (in dev version 0.1.2.5, and CRAN version 0.1.3).

> It is a bit unclear to me how (or whether) nice_contrasts() can handle nested comparisons for marginal means. In the syntax of emmeans, it is straightforward to define whether you want to know contrasts for levels of var1 within var2, and vice versa.

Thanks, I have added clarification in the function documentation:

``` r
#' Does not for the moment support nested comparisons for marginal means,
#' only a comparison of all groups. For nested comparisons, please use
#' [emmeans::contrast()] directly, or for the *easystats* equivalent,
#' [modelbased::estimate_contrasts()].
```

---

Finally, would you have any comments or suggestions on the following points:

- Content, structure, or organization of the paper
- Names of the functions (based on conventions, etc.)
- Names of the arguments within the functions (based on conventions, etc.)

Thanks again.

**Comment by avakiai** [comment]
Hi Remi, 
Apologies for my late reply. This looks great! Thank you for implementing those changes. 

I have no further comments - naming conventions and the paper seem ok to me. Vignettes are also well-written. 

I will now check those boxes I had left unchecked in my inital review. 

Best wishes,
Ava
